pkg/actor/upgrade.go

```diff
@@ -0,0 +1,192 @@
+
+	stsName := cluster.StatefulSetName()
+
+	key := kubetypes.NamespacedName{
+		Namespace: cluster.Namespace(),
+		Name:      stsName,
+	}
+	ss := &appsv1.StatefulSet{}
+	if err := up.client.Get(ctx, key, ss); err != nil {
+		return errors.Wrap(err, "failed to fetch statefulset")
+	}
+
+	if statefulSetIsUpdating(ss) {
+		return NotReadyErr{Err: errors.New("statefulset is updating, waiting for the update to finish")}
+	}
+
+	// Get version from spec
+	specStr := getVersionFromImage(cluster.Spec().Image)
+	if specStr == "" {
+		return fmt.Errorf("Unknown CockroachDB version in spec: %s", cluster.Spec().Image)
+	}
+
+	specVer, err := semver.NewVersion(specStr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse spec image version: %s", specStr)
+	}
+
+	clusterStr, err := up.getClusterSetting(cluster, "version")
+	if err != nil {
+		return errors.Wrapf(err, "failed to fetch cluster version")
+	}
+
+	contraintStr := fmt.Sprintf("^%s", clusterStr)
+	constraint, err := semver.NewConstraint(contraintStr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to initialize version contaraint: %s", contraintStr)
+	}
+
+	log.Info("detected versions", "spec", specVer.String(), "cluster", clusterStr)
+
+	// Major upgrade
+	if !constraint.Check(specVer) {
+		log.Info("preserving downgrade option")
+		if err := up.preserveDowngrade(cluster, clusterStr); err != nil {
+			return err
+		}
+	}
+
+	dbContainer, err := kube.FindContainer(resource.DbContainerName, &ss.Spec.Template.Spec)
+	if err != nil {
+		return err
+	}
+
+	// nothing to be done
+	if dbContainer.Image == cluster.Spec().Image {
+		log.Info("no version changes are required")
+		return nil
+	}
+
+	update := ss.DeepCopy()
+	updatedContainer, _ := kube.FindContainer(resource.DbContainerName, &update.Spec.Template.Spec)
+	updatedContainer.Image = cluster.Spec().Image
+
+	if err := up.client.Update(ctx, update); err != nil {
+		return errors.Wrap(err, "failed to update image version")
+	}
+
+	log.Info("scheduled update", "new version", specStr)
+	CancelLoop(ctx)
+	return nil
+}
+
+func getVersionFromImage(image string) string {
+	i := strings.LastIndex(image, ":")
+	if i == -1 {
+		return ""
+	}
+
+	return image[i+1:]
+}
+
+func (up *upgrade) getClusterSetting(cluster *resource.Cluster, setting string) (string, error) {
+	rows, err := up.showClusterSetting(cluster, setting)
+	if err != nil {
+		return "", err
+	}
+
+	if len(rows) == 1 {
+		// we only have the header.
+		return "", nil
+	}
+
+	if len(rows) != 2 || len(rows[0]) != 1 || len(rows[1]) != 1 {
+		return "", fmt.Errorf("expected two rows with 1 column, got: %v", rows)
+	}
+
+	if rows[0][0] != setting {
+		return "", fmt.Errorf("expected column named \"%s\", got: %s", setting, rows[0][0])
+	}
+
+	return rows[1][0], nil
+}
+
+func (up *upgrade) showClusterSetting(cluster *resource.Cluster, setting string) ([][]string, error) {
+	cmd := []string{
+		"/bin/bash",
+		"-c",
+		"/cockroach/cockroach sql " + cluster.SecureMode() +
```

Comment:
How does this work in secure mode? Are certs sitting in the operator container? Which certs & how does this happen?

Reply:
Yep, the certificates exist in all pods in secure mode.